xbean-finder/src/main/java/org/apache/xbean/finder/archive/JarArchive.java

```diff
@@ -41,22 +43,52 @@ public class JarArchive implements Archive {
     private final JarFile jar;
     private final MJarSupport mjar = new MJarSupport();
 
-    public JarArchive(ClassLoader loader, URL url) {
+    public JarArchive(ClassLoader loader, URL url){
 //        if (!"jar".equals(url.getProtocol())) throw new IllegalArgumentException("not a jar url: " + url);
 
+        this.loader = loader;
+        this.url = url;
+        File jf;
+        int idx;
+
+        /*
+         * True URL.getPath(...) wiping out protocol prefixes:
+         *
+         *  1. 'jar:file:/c:/temp/...' -> 'jar' + 'file:/c:/temp/...'
+         *  2. 'file:/c:/temp/...' -> 'file' + '/c:/temp/...'
+         *
+         *  assuming we accept 'file:/...${xxx.jar}!/' URLs
+         *      AND 'zip:file:/...!/' would be cool too, if
+         *      allowed by new URL(...)
+         */
         try {
-            this.loader = loader;
-            this.url = url;
-            URL u = url;
-
-            String jarPath = url.getFile();
-            if (jarPath.contains("!")) {
-                jarPath = jarPath.substring(0, jarPath.lastIndexOf("!"));
-                u = new URL(jarPath);
+            // Underlying JarFile(...) requires RandomAccessFile,
+            //  so only local file URLs here...
+            while(!"file".equalsIgnoreCase(url.getProtocol())) {
```

Comment:
looks too relaxed, `jar:foo:file` wouldn't fail but it should using this new implementation.

Reply:
jar:foo:file is actually prohibited by new URL(), why not allow this impossible scenario?
from other side, if they let e.g. tar:gz:file, this will work here and, if fail - deeper, where e.g. 'tar' format should be supported. 

in general, AFAIK we are interested here in url, pointing to (available, because of possible '!'s in path) local zip file: '...file:/.../some.jar....'. i think:
1. a File parameter, instead of URL would be appropriate
2. extra prefixes and suffixes may cause warning, but shouldn't ruin jar access.

Reply:
> > jar:foo:file is actually prohibited by new URL(), why not allow this impossible scenario?
> 
> Prohibited by who? Thing is that we actually only support `jar` and `file`, now you allow `jar:(whatever:)*file` - this is allowed by `URL` and `URLHandler` API. Don't think it is what we want so we should just unwrap once or let the caller unwrap it more before passing the path to the archive IMHO.
i cannot do e.g. new URL("jar:foa:file:/c:/Temp"), but - wow - can new URL("http:rtsp:ftp:jar:file:/c:/Temp")!
that's technically bullshit, but you're right - possible. 

btw, absolhuck-o-lutely have no idea why don't URL.getProtocol() just say once "jar:file" instead of nesting URLs. or, if nesting, not to be e.g. "file:/usr/lib.jar:jar:/inner/file.txt"!
> Nested jar support is done through other `Archive` implementation like https://github.com/Talend/component-runtime/blob/master/component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/xbean/NestedJarArchive.java but it requires some knowledge about the enclosing structure so better to not try to be too clever there IMHO.
yep, too clever with restrictions as well ;) can't shake the feeling that this is what I'm doing...

Reply:
Thing is you allow other urls than "jar:file" with this loop so all the mentionned cases.

Side note: foo can really be used, just requires to register a handler. Famous example: pax-url and mvn handler.